File: src/main/java/me/miki/shindo/management/remote/changelog/ChangelogManager.java
package me.miki.shindo.management.remote.changelog;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import me.miki.shindo.Shindo;
import me.miki.shindo.utils.JsonUtils;
import me.miki.shindo.utils.Multithreading;
import me.miki.shindo.utils.network.HttpUtils;

import java.util.concurrent.CopyOnWriteArrayList;

public class ChangelogManager {

    private final CopyOnWriteArrayList<Changelog> changelogs = new CopyOnWriteArrayList<Changelog>();

    public ChangelogManager() {
        Multithreading.runAsync(() -> loadChangelog());
    }

    private void loadChangelog() {

        JsonObject jsonObject = HttpUtils.readJson("https://cdn.shindoclient.com/data/changelogs/versions/" + Shindo.getInstance().getVerIdentifier() + ".json", null);


        if (jsonObject != null) {

            JsonArray jsonArray = JsonUtils.getArrayProperty(jsonObject, "changelogs");

            if (jsonArray != null) {

                for (JsonElement jsonElement : jsonArray) {

                    Gson gson = new Gson();
                    JsonObject changelogJsonObject = gson.fromJson(jsonElement, JsonObject.class);

                    changelogs.add(new Changelog(JsonUtils.getStringProperty(changelogJsonObject, "text", "null"),
                            ChangelogType.getTypeById(JsonUtils.getIntProperty(changelogJsonObject, "type", 999))));
                }
            }
        }
    }

    public CopyOnWriteArrayList<Changelog> getChangelogs() {
        return changelogs;
    }
}

File: src/main/kotlin/me/miki/shindo/management/remote/changelog/ChangelogManager.kt
package me.miki.shindo.management.remote.changelog

import com.google.gson.Gson
import com.google.gson.JsonObject
import me.miki.shindo.Shindo
import me.miki.shindo.utils.JsonUtils
import me.miki.shindo.utils.concurrent.TaskExecutor
import me.miki.shindo.utils.concurrent.ThreadPoolType
import me.miki.shindo.utils.network.HttpUtils
import java.util.concurrent.CopyOnWriteArrayList

class ChangelogManager {

    private val changelogs = CopyOnWriteArrayList<Changelog>()

    init {
        TaskExecutor.runAsync(ThreadPoolType.NETWORK) { loadChangelog() }
    }

    private fun loadChangelog() {
        val jsonObject = HttpUtils.readJson(
            "https://cdn.shindoclient.com/data/changelogs/versions/${Shindo.getInstance().verIdentifier}.json",
            null
        ) ?: return
        val jsonArray = JsonUtils.getArrayProperty(jsonObject, "changelogs") ?: return
        val gson = Gson()
        for (jsonElement in jsonArray) {
            val changelogJsonObject = gson.fromJson(jsonElement, JsonObject::class.java)
            changelogs.add(
                Changelog(
                    JsonUtils.getStringProperty(changelogJsonObject, "text", "null").toString(),
                    ChangelogType.getTypeById(JsonUtils.getIntProperty(changelogJsonObject, "type", 999))
                )
            )
        }
    }

    fun getChangelogs(): CopyOnWriteArrayList<Changelog> = changelogs
}
